vpr/src/pack/sync_netlists_to_routing_flat.cpp

```diff
@@ -0,0 +1,367 @@
+            int sink_pin = rr_graph.node_pin_num(sink_inode);
+
+            auto [_, subtile] = get_sub_tile_from_pin_physical_num(physical_tile, source_pin);
+
+            ClusterBlockId clb = placement_ctx.grid_blocks.block_at_location({
+                rr_graph.node_xlow(source_inode),
+                rr_graph.node_ylow(source_inode),
+                subtile,
+                rr_graph.node_layer(source_inode)
+            });
+
+            /* Look up pb graph pins from pb type if pin is not on tile, look up from block otherwise */
+            const t_pb_graph_pin* source_pb_graph_pin, *sink_pb_graph_pin;
+            if(is_pin_on_tile(physical_tile, sink_pin)){
+                sink_pb_graph_pin = get_pb_graph_node_pin_from_block_pin(clb, sink_pin);
+            }else{
+                sink_pb_graph_pin = get_pb_pin_from_pin_physical_num(physical_tile, sink_pin);
+            }
+            if(is_pin_on_tile(physical_tile, source_pin)){
+                source_pb_graph_pin = get_pb_graph_node_pin_from_block_pin(clb, source_pin);
+            }else{
+                source_pb_graph_pin = get_pb_pin_from_pin_physical_num(physical_tile, source_pin);
+            }
+
+            t_pb* pb = cluster_ctx.clb_nlist.block_pb(clb);
+
+            /* Route between the pins */
+            route_intra_cluster_conn(source_pb_graph_pin, sink_pb_graph_pin, convert_to_atom_net_id(net_id), pb);
+        }
+    }
+}
+
+/** Rebuild the ClusterNetId <-> AtomNetId lookup after compressing the ClusterNetlist
+ * Needs the "most recent" ClusterNetIds in atom_ctx.lookup: won't work after invalidating the ClusterNetIds twice */
+inline void rebuild_atom_nets_lookup(ClusteredNetlist::IdRemapper& remapped){
+    auto& atom_ctx = g_vpr_ctx.mutable_atom();
+    auto& atom_lookup = atom_ctx.lookup;
+
+    for(auto parent_net_id: atom_ctx.nlist.nets()){
+        auto atom_net_id = convert_to_atom_net_id(parent_net_id);
+        ClusterNetId old_clb_net = atom_lookup.clb_net(atom_net_id);
+        if(!old_clb_net)
+            continue;
+        ClusterNetId new_clb_net = remapped.new_net_id(old_clb_net);
+        atom_lookup.set_atom_clb_net(atom_net_id, new_clb_net);
+    }
+}
+
+/** Regenerate clustered netlist nets from routing results */
+static void sync_clustered_netlist_to_routing(void){
+    auto& cluster_ctx = g_vpr_ctx.mutable_clustering();
+    auto& place_ctx = g_vpr_ctx.mutable_placement();
+    auto& route_ctx = g_vpr_ctx.routing();
+    auto& clb_netlist = cluster_ctx.clb_nlist;
+    auto& device_ctx = g_vpr_ctx.device();
+    auto& rr_graph = device_ctx.rr_graph;
+    auto& atom_ctx = g_vpr_ctx.mutable_atom();
+    auto& atom_lookup = atom_ctx.lookup;
+
+    /* 1. Remove all nets, pins and ports from the clustered netlist (except clocks) */
+    for(auto net_id: clb_netlist.nets()){
+        auto atom_net_id = atom_lookup.atom_net(net_id);
+        if(route_ctx.is_clock_net[atom_net_id])
```

Comment:
Comment / assert ideal clock.

Reply:
Or better yet, update this code to "clock net is not in the netlist" somehow. Then this code is ready for the non-ideal clock case.

Reply:
done